pkg/setup/config.go

```diff
@@ -48,19 +61,248 @@ func InitConfig(vConfig *config.VirtualClusterConfig) error {
 		translate.Default = translate.NewSingleNamespaceTranslator(vConfig.WorkloadTargetNamespace)
 	}
 
+	if err := EnsureBackingStoreChanges(
+		ctx,
+		vConfig.ControlPlaneClient,
+		vConfig.Name,
+		vConfig.ControlPlaneNamespace,
+		vConfig.Distro(),
+		vConfig.BackingStoreType(),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// EnsureBackingStoreChanges ensures that only a certain set of allowed changes to the backing store and distro occur.
+func EnsureBackingStoreChanges(ctx context.Context, client kubernetes.Interface, name, namespace, distro string, backingStoreType vclusterconfig.StoreType) error {
+	if ok, err := CheckUsingHelm(ctx, client, name, namespace, distro, backingStoreType); err != nil {
+		return err
+	} else if ok {
+		return nil
+	}
+
+	if ok, err := CheckUsingSecretAnnotation(ctx, client, name, namespace, distro, backingStoreType); err != nil {
+		return fmt.Errorf("using secret annotations: %w", err)
+	} else if ok {
+		if err := updateSecretAnnotations(ctx, client, name, namespace, distro, backingStoreType); err != nil {
+			return fmt.Errorf("update secret annotations: %w", err)
+		}
+
+		return nil
+	}
+
+	if ok, err := CheckUsingHeuristic(distro); err != nil {
+		return fmt.Errorf("using heuristic: %w", err)
+	} else if ok {
+		if err := updateSecretAnnotations(ctx, client, name, namespace, distro, backingStoreType); err != nil {
+			return fmt.Errorf("update secret annotations: %w", err)
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+// CheckUsingHelm fetches the previous release revision and its computed values, and then reconstructs the distro and storage settings.
+func CheckUsingHelm(ctx context.Context, client kubernetes.Interface, name, namespace, distro string, backingStoreType vclusterconfig.StoreType) (bool, error) {
+	ls := kblabels.Set{}
+	ls["name"] = name
+	releases, err := helm.NewSecrets(client).ListUnfiltered(ctx, ls.AsSelector(), namespace)
+	if err != nil || len(releases) == 0 {
+		return false, nil
+	}
+
+	// (ThomasK33): if there is only one revision, we're dealing with an initial installation
+	// at which point we can just exit
+	if len(releases) == 1 {
```

Comment:
Why is that? Couldn't it be that a user only installed one 0.19.x revision before?

Reply:
No since v0.20 code is executed as this is part of the control plane image, so it must be v0.20 as we know latest release is v0.20.